### PR TITLE
add step remote upload for iso builder

### PR DESCRIPTION
**Pull request by remijouannet**
this PR include a step to upload an iso to vpshere for the builder vsphere-iso

I used the generic iso downloader from packer to include it in vpshere-iso
the limitation here is you can't use iso_paths and iso_urls at the same time, because iso_urls overwrite the parameter iso_paths after the iso is upload

the ISO location on the datastore is hardcode to ISO/, the folder is create if not exists

(I misclicked and close https://github.com/jetbrains-infra/packer-builder-vsphere/pull/146 by mistake)

### CODE REVIEW

**Comment by mkuzmin** [comment]
related issue #99

**Comment by remijouannet** [comment]
to answer the same question

    what datastore should the file be uploaded to?
        I use the "datastore" parameter that's already include in this plugins

    delete it after a build or not?
        I don't for the moment, i can add a parameter for that if it's really needed

    how to share the same files between different Packer builds?
        since I can't make a checksum on the datastore, the code only check the filename

    will you use such option in your daily production builds or only in demos?
        For production I'm  working on a jenkins pipeline right now for that

**Comment by remijouannet** [comment]
@mkuzmin let me know is there anything wrong with that patch, 
I'm currently using a fork of your plugin in production but i'm trying to push every patch

**Comment by mkuzmin** [comment]
I'm sorry that took so much time on my side again.
If you are still interested in finishing the feature, please rebase it onto master, so PR has just a single commit. Otherwise, I can do it myself.

**Comment by remijouannet** [comment]
No problem, I will rebase it tonight or tomorrow

File: common/step_remote_upload.go [hunk 1]
```diff
@@ -0,0 +1,59 @@
...
+
+func (s *StepRemoteUpload) Run(_ context.Context, state multistep.StateBag) multistep.StepAction {
+	ui := state.Get("ui").(packer.Ui)
+	d := state.Get("driver").(*driver.Driver)
+
+	if path, ok := state.GetOk(s.Path); ok {
+		filename := filepath.Base(path.(string))
+		remotepath := fmt.Sprintf("ISO/%s", filename)
+		remotedirectory := fmt.Sprintf("[%s] ISO/", s.Datastore)
+		final_remotepath := fmt.Sprintf("[%s] %s", s.Datastore, remotepath)
```
Comment (mkuzmin) [inline common/step_remote_upload.go]:
```suggestion
 		final_remotepath := fmt.Sprintf("%s/%s", remotedirectory, remotepath)
```

File: common/step_remote_upload.go [hunk 2]
```diff
@@ -0,0 +1,59 @@
...
+
+		ui.Say(fmt.Sprintf("Uploading %s to %s", filename, remotepath))
+
+		ds, err := d.FindDatastore(s.Datastore, s.Host)
+		if err != nil {
+			ui.Say("Datastore doesn't exist")
+			state.Put("error", err)
```
Comment (mkuzmin) [inline common/step_remote_upload.go]:
```suggestion
 			state.Put("error", fmt.Errorf("datastore doesn't exist: %v",err))
```
full error message into STDERR, not to `ui.Say`

File: common/step_remote_upload.go [hunk 3]
```diff
@@ -0,0 +1,59 @@
+package common
```
Comment (mkuzmin) [inline common/step_remote_upload.go]: CD-ROM logic is not used in `clone` builder. let's keep it in iso package

File: iso/step_add_cdrom.go [hunk 4]
```diff
@@ -38,13 +39,20 @@ func (s *StepAddCDRom) Run(_ context.Context, state multistep.StateBag) multiste
...
+  
 	ui.Say("Mount ISO images...")
-	for _, path := range s.Config.ISOPaths {
-		if err := vm.AddCdrom(s.Config.CdromType, path); err != nil {
+	if path, ok := state.GetOk(s.RemotePath); ok {
+		if err := vm.AddCdrom(path.(string)); err != nil {
 			state.Put("error", fmt.Errorf("error mounting an image: %v", err))
 			return multistep.ActionHalt
 		}
+	} else {
```
Comment (mkuzmin) [inline iso/step_add_cdrom.go]: we need to reverse the order. existing images are mounted first (like OS distro), dynamically uploaded image later

**Comment by mkuzmin** [comment]
The branch has a set of merges from 'master', so rebasing is quite complicated.
Here is a good post https://seesparkbox.com/foundry/how_to_not_dread_rebases_when_managing_long_lived_feature_branches

You need to do something like 
```
git checkout 4f952613
git checkout -b upload
git diff HEAD pull/164 >a.diff
git apply a.diff 
git commit
```

File: common/step_remote_upload.go [hunk 5]
```diff
@@ -0,0 +1,59 @@
...
+}
+
+func (s *StepRemoteUpload) Run(_ context.Context, state multistep.StateBag) multistep.StepAction {
+	ui := state.Get("ui").(packer.Ui)
+	d := state.Get("driver").(*driver.Driver)
+
+	if path, ok := state.GetOk(s.Path); ok {
+		filename := filepath.Base(path.(string))
+		remotepath := fmt.Sprintf("ISO/%s", filename)
+		remotedirectory := fmt.Sprintf("[%s] ISO/", s.Datastore)
```
[the remark this answers is not in the document]
Comment (remijouannet) [reply common/step_remote_upload.go]: I'm not sure i understand this one, do you want to handle remote upload even if the user doesn't specify datastore ?

**Comment by remijouannet** [comment]
Hi @mkuzmin, I rebased everything and push just one commit, I "almost" solved all comments, tell me if this is OK now

File: common/step_remote_upload.go (hunk 5 above)
[the remark this answers is not in the document]
Comment (mkuzmin) [reply common/step_remote_upload.go]: > handle remote upload even if the user doesn't specify datastore ?

exactly. if a host has a single attached datastore, we can detect it automatically.

you make FindDatastore() in next lines, just reverse the order.

[the remark this answers is not in the document]
Comment (remijouannet) [reply common/step_remote_upload.go]: ho all right, i just do FindDatastore and if the datastore parameter is empty I just take the first datastore that the host have

[the remark this answers is not in the document]
Comment (mkuzmin) [reply common/step_remote_upload.go]: you call `FindDatastore("","esxi-1")`.
If the host has just single attached datastore then it's automatically resolved.
In the host has multiple datastores - that's an error. the user has to provide it explicitly.

[the remark this answers is not in the document]
Comment (remijouannet) [reply common/step_remote_upload.go]: should be good now

**Comment by remijouannet** [comment]
done

**Comment by mkuzmin** [comment]
Thank you!

**Comment by remijouannet** [comment]
👍

**Comment by mkuzmin** [comment]
I've tested a mixed set of pre-created and dynamically uploaded ISO images for macOS build (4586e8d3b666afad8e00a3355fea6e5828b2c9f9), and it actually works!

**Comment by remijouannet** [comment]
glad to hear it :)

**Comment by RPitt** [comment]
Great stuff @mkuzmin 

However looking at this it seems to be restricted to uploading/mounting of a single ISO (albeit from multiple sources) ... but doesn't facilitate uploading 2 different ISOs (e.g. Windows setup.ISO + VMWare tools). OR have I misread things?

Thanks in advance.

**Comment by mkuzmin** [comment]
That’s correct, multiple URLs for the same image. 
This is how the core Packer feature is designed. Inside a plug-in we don’t have any flexibility. 

What’s the use case?

**Comment by RPitt** [comment]
Understood, a packer limitation.
The use case I had in mind was VM creation from Windows ISO + VMWare Tools ISO, pretty much like your Windows example.

**Pull request merged**